subsys/net/l2/ethernet/ptp/ptp_master.c

```diff
@@ -0,0 +1,400 @@
+
+	/* enqueue for sending */
+	net_if_queue_tx(state.iface, pkt);
+}
+
+static void handle_ptp_delay_req(struct net_pkt *req_pkt, struct ptp_header *req_ptp_hdr)
+{
+	struct net_ptp_time *ts = net_pkt_timestamp(req_pkt);
+
+	/* prepare body */
+	struct ptp_delay_resp_body ptp_body;
+
+	memset(&ptp_body, 0, sizeof(ptp_body));
+	ptp_ts_net_to_wire(&ptp_body.receive_timestamp, ts);
+	memcpy(&ptp_body.requesting_port_identity, &req_ptp_hdr->source_port_identity,
+		sizeof(ptp_body.requesting_port_identity));
+
+	/* prepare header */
+	struct ptp_header ptp_hdr;
+
+	memset(&ptp_hdr, 0, sizeof(ptp_hdr));
+	ptp_hdr.message_type = PTP_MESSAGE_TYPE_DELAY_RESP;
+	ptp_hdr.version_ptp = 2;
+	ptp_hdr.minor_version_ptp = 0;
+	ptp_hdr.message_length = sizeof(ptp_hdr) + sizeof(ptp_body);
+	ptp_hdr.domain_number = state.domain_number;
+	memcpy(&ptp_hdr.source_port_identity, &state.port_identity,
+		sizeof(ptp_hdr.source_port_identity));
+	ptp_hdr.sequence_id = req_ptp_hdr->sequence_id;
+	ptp_hdr.log_message_interval = state.log_min_delay_req_interval;
+
+	/* allocate packet with PTP settings */
+	struct net_pkt *pkt = net_pkt_alloc_with_buffer(state.iface, ptp_hdr.message_length,
+		AF_UNSPEC, 0, K_NO_WAIT);
+
+	net_pkt_set_ptp(pkt, true);
+	net_pkt_set_priority(pkt, NET_PRIORITY_CA);
+
+	/* set src mac addr */
+	struct net_linkaddr *iface_addr = net_if_get_link_addr(state.iface);
+
+	net_pkt_lladdr_src(pkt)->addr = iface_addr->addr;
+	net_pkt_lladdr_src(pkt)->len = iface_addr->len;
+
+	/* set dst mac addr */
+	net_pkt_lladdr_dst(pkt)->addr = (uint8_t *)&ptp_mcast_eth_addr;
+	net_pkt_lladdr_dst(pkt)->len = sizeof(ptp_mcast_eth_addr);
+
+	/* write packet PTP data */
+	net_pkt_write(pkt, &ptp_hdr, sizeof(ptp_hdr));
+	net_pkt_write(pkt, &ptp_body, sizeof(ptp_body));
+
+	/* enqueue for sending */
+	net_if_queue_tx(state.iface, pkt);
+}
+
+static void pkt_received(struct net_context *context, struct net_pkt *pkt,
+			 union net_ip_header *ip_hdr, union net_proto_header *proto_hdr,
+			 int status, void *user_data)
+{
+	if (ntohs(NET_ETH_HDR(pkt)->type) != NET_ETH_PTYPE_PTP) {
+		goto done;
+	}
+
+	/* skip ethernet header */
+	net_pkt_skip(pkt, sizeof(struct net_eth_hdr));
+
+	/* read PTP header */
+	struct ptp_header ptp_hdr;
+
+	net_pkt_read(pkt, &ptp_hdr, sizeof(ptp_hdr));
+
+	if (ptp_hdr.version_ptp != 2 || ptp_hdr.minor_version_ptp != 0) {
+		LOG_WRN("Unsupported PTP v%d.%d packet ignored", ptp_hdr.version_ptp,
+			ptp_hdr.minor_version_ptp);
+		goto done;
+	}
+
+	switch (ptp_hdr.message_type) {
+	case PTP_MESSAGE_TYPE_DELAY_REQ:
+		handle_ptp_delay_req(pkt, &ptp_hdr);
+		break;
+	}
+
+done:
+	net_pkt_unref(pkt);
+}
+
+static void ptp_announce_sender(void)
+{
+	for (;;) {
+		if (state.running) {
+			send_ptp_announce();
+		}
+		k_msleep(1000 * (1 << state.log_announce_interval));
+	}
+}
+
+static void ptp_sync_sender(void)
+{
+	for (;;) {
+		if (state.running) {
+			send_ptp_sync();
+		}
+		k_msleep(1000 * (1 << state.log_sync_interval));
+	}
+}
```

Comment:
I think that instead of allocating memory for two extra threads, which only purpose is to send a packet periodically, using delayed work would be a much better option (https://docs.zephyrproject.org/2.6.0/reference/kernel/threads/workqueue.html#delayable-work).

Instead of using `k_msleep()` to wait, with delayed work you can specify when should it be executed. Seeing that `announce()/sync()` are not blocking calls, using a default system work queue should be fine here.